### PR TITLE
Create SCCNonlinearSolve

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Requires https://github.com/SciML/SciMLBase.jl/pull/861

```julia
using NonlinearSolve, SCCNonlinearSolve

function f(du,u,p)
	du[1] = cos(u[2]) - u[1]
	du[2] = sin(u[1] + u[2]) + u[2]
	du[3] = 2u[4] + u[3] + 1.0
	du[4] = u[5]^2 + u[4]
	du[5] = u[3]^2 + u[5]
	du[6] = u[1] + u[2] + u[3] + u[4] + u[5]    + 2.0u[6] + 2.5u[7] + 1.5u[8]
	du[7] = u[1] + u[2] + u[3] + 2.0u[4] + u[5] + 4.0u[6] - 1.5u[7] + 1.5u[8]
	du[8] = u[1] + 2.0u[2] + 3.0u[3] + 5.0u[4] + 6.0u[5] + u[6] - u[7] - u[8]
end
prob = NonlinearProblem(f, zeros(8))
sol = solve(prob)

u0 = zeros(2)
p = zeros(3)

function f1(du,u,p)
	du[1] = cos(u[2]) - u[1]
	du[2] = sin(u[1] + u[2]) + u[2]
end
explicitfun1(p,sols) = nothing
prob1 = NonlinearProblem(NonlinearFunction{true, SciMLBase.NoSpecialize}(f1), zeros(2), p)
sol1 = solve(prob1, NewtonRaphson())

function f2(du,u,p)
	du[1] = 2u[2] + u[1] + 1.0
	du[2] = u[3]^2 + u[2]
	du[3] = u[1]^2 + u[3]
end
explicitfun2(p,sols) = nothing
prob2 = NonlinearProblem(NonlinearFunction{true, SciMLBase.NoSpecialize}(f2), zeros(3), p)
sol2 = solve(prob2, NewtonRaphson())

function f3(du,u,p)
	du[1] = p[1] + 2.0u[1] + 2.5u[2] + 1.5u[3]
	du[2] = p[2] + 4.0u[1] - 1.5u[2] + 1.5u[3]
	du[3] = p[3] + + u[1] - u[2] - u[3]
end
prob3 = NonlinearProblem(NonlinearFunction{true, SciMLBase.NoSpecialize}(f3), zeros(3), p)
function explicitfun3(p,sols)
	p[1] = sols[1][1] + sols[1][2] + sols[2][1] + sols[2][2] + sols[2][3]
	p[2] = sols[1][1] + sols[1][2] + sols[2][1] + 2.0sols[2][2] + sols[2][3]
	p[3] = sols[1][1] + 2.0sols[1][2] + 3.0sols[2][1] + 5.0sols[2][2] + 6.0sols[2][3]
end
explicitfun3(p,[sol1,sol2])
sol3 = solve(prob3, NewtonRaphson())
manualscc = [sol1; sol2; sol3]

sccprob = SciMLBase.SCCNonlinearProblem([prob1,prob2,prob3], SciMLBase.Void{Any}.([explicitfun1,explicitfun2,explicitfun3]))
scc_sol = solve(sccprob, NewtonRaphson())
sol ≈ manualscc ≈ scc_sol
```

Right now this just handles `solve`. I want to get a first version in, then the todos for next steps are:

* `init`
* `step!`
* Put appropriate trait calculations on this

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Note the `SciMLBase.Void{Any}.([explicitfun1,explicitfun2,explicitfun3])` can likely be done internally in the SCCNonlinearProblem constructor. This also requires `SciMLBase.NoSpecialize` which we are going to make the default after a few things from @oscardssmith first.

This will start as undocumented and experimental so we can work out the kinks.

It has a few allocations, I'm not worrying about those yet, correctness first.

File: lib/SCCNonlinearSolve/test/core_tests.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            _sol = SciMLBase.build_solution(prob.probs[i], nothing, sol.u, sol.resid, retcode = sol.retcode)
+            sols[i] = _sol
+        end
+        
+        # TODO: fix allocations with a lazy concatenation
+        u .= reduce(vcat,sols)
+        resid .= reduce(vcat,getproperty.(sols,:resid))
+
+        SciMLBase.build_solution(prob, alg, u, resid, original = sols)
+    end
```
Comment (AayushSabharwal) [inline lib/SCCNonlinearSolve/test/core_tests.jl:74]:
```suggestion
```

File: lib/SCCNonlinearSolve/test/core_tests.jl [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    sccprob = SciMLBase.SCCNonlinearProblem([prob1,prob2,prob3], SciMLBase.Void{Any}.([explicitfun1,explicitfun2,explicitfun3]))
```
Comment (AayushSabharwal) [inline lib/SCCNonlinearSolve/test/core_tests.jl:76]: Just out of curiosity, what is the advantage of `Vector{SciMLBase.Void{Any}}` as opposed to `Vector{Function}`?

File: lib/SCCNonlinearSolve/test/core_tests.jl [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    scc_sol = solve(sccprob, NewtonRaphson())
+    sol ≈ manualscc ≈ scc_sol
```
Comment (AayushSabharwal) [inline lib/SCCNonlinearSolve/test/core_tests.jl:78]:
```suggestion
    @test sol ≈ manualscc ≈ scc_sol
```

**Review by AayushSabharwal — COMMENTED** [review]
Looks really cool! Mostly just general readability/interface comments. I'm looking forward to targeting this with MTK, not so much figuring out how SII will have to interact with it 😅

File: lib/SCCNonlinearSolve/src/SCCNonlinearSolve.jl [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+function SciMLBase.solve(prob::SciMLBase.SCCNonlinearProblem, alg; kwargs...)
+	numscc = length(prob.probs)
+	sols = [SciMLBase.build_solution(prob, nothing, prob.u0, convert(eltype(prob.u0),NaN)*prob.u0) for prob in prob.probs]
+	u = reduce(vcat,[prob.u0 for prob in prob.probs])
+	resid = copy(u)
+
+    earlyexit = false
+    lasti = 1
+	for i in 1:numscc
+		prob.explictfuns![i](prob.probs[i].p[1],sols)
```
[the remark this answers is not in the document]
Comment (AayushSabharwal) [reply lib/SCCNonlinearSolve/src/SCCNonlinearSolve.jl:14]: @ChrisRackauckas

File: lib/SCCNonlinearSolve/test/core_tests.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply lib/SCCNonlinearSolve/test/core_tests.jl:76]: Type stability and performance optimizations under the assumption that p is not isbits.

**Comment by ChrisRackauckas** [comment]
@AayushSabharwal I see where the confusion was from yesterday: I didn't have the fully final version pushed to the package. This is now set to the parameter mutation form I mentioned.

File: lib/SCCNonlinearSolve/test/core_tests.jl [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        du[2] = sin(u[1] + u[2]) + u[2]
+        du[3] = 2u[4] + u[3] + 1.0
+        du[4] = u[5]^2 + u[4]
+        du[5] = u[3]^2 + u[5]
+        du[6] = u[1] + u[2] + u[3] + u[4] + u[5] + 2.0u[6] + 2.5u[7] + 1.5u[8]
+        du[7] = u[1] + u[2] + u[3] + 2.0u[4] + u[5] + 4.0u[6] - 1.5u[7] + 1.5u[8]
+        du[8] = u[1] + 2.0u[2] + 3.0u[3] + 5.0u[4] + 6.0u[5] + u[6] - u[7] - u[8]
+    end
+    prob = NonlinearProblem(f, zeros(8))
+    sol = solve(prob)
```
Comment (ChrisRackauckas) [inline lib/SCCNonlinearSolve/test/core_tests.jl:20]:
```suggestion
    sol = solve(prob, NewtonRaphson())
```

File: lib/SCCNonlinearSolve/test/qa_tests.jl [hunk 6]
```diff
@@ -0,0 +1,22 @@
+@testitem "Aqua" tags=[:core] begin
+    using Aqua, SCCNonlinearSolve
+
+    Aqua.test_all(
+        SCCNonlinearSolve;
+        piracies = false, ambiguities = false, stale_deps = false, deps_compat = false
+    )
+    Aqua.test_stale_deps(SCCNonlinearSolve; ignore = [:SciMLJacobianOperators])
+    Aqua.test_deps_compat(SCCNonlinearSolve; ignore = [:SciMLJacobianOperators])
+    Aqua.test_piracies(SCCNonlinearSolve)
```
Comment (ChrisRackauckas) [inline lib/SCCNonlinearSolve/test/qa_tests.jl:10]:
```suggestion
    Aqua.test_piracies(SCCNonlinearSolve; treat_as_own = [SCCNonlinearSolve.SciMLBase.solve])
```

File: lib/SCCNonlinearSolve/test/qa_tests.jl [hunk 7]
```diff
@@ -0,0 +1,23 @@
+@testitem "Aqua" tags=[:core] begin
+    using Aqua, SCCNonlinearSolve
+
+    Aqua.test_all(
+        SCCNonlinearSolve;
+        piracies = false, ambiguities = false, stale_deps = false, deps_compat = false
+    )
+    Aqua.test_stale_deps(SCCNonlinearSolve; ignore = [:SciMLJacobianOperators, :NonlinearSolveBase])
+    Aqua.test_deps_compat(SCCNonlinearSolve; ignore = [:SciMLJacobianOperators])
```
Comment (ChrisRackauckas) [inline lib/SCCNonlinearSolve/test/qa_tests.jl:9]:
```suggestion
    Aqua.test_deps_compat(SCCNonlinearSolve; ignore = [:SciMLJacobianOperators, :NonlinearSolveBase])
```

**Pull request merged**